File: app/src/main/java/itis/ru/justtalk/adapters/people/CardPagerAdapterS.java
package itis.ru.justtalk.adapters.people;

/**
 * Created by KottlandPro TET on 3/3/2018.
 */

import android.graphics.drawable.Drawable;
import android.support.v4.view.PagerAdapter;
import android.support.v7.widget.CardView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.RequestBuilder;
import com.bumptech.glide.load.Transformation;
import com.bumptech.glide.request.RequestOptions;

import java.util.ArrayList;
import java.util.List;

import itis.ru.justtalk.R;
import itis.ru.justtalk.models.User;
import jp.wasabeef.glide.transformations.RoundedCornersTransformation;

public class CardPagerAdapterS extends PagerAdapter implements CardAdapter {

    private List<CardView> mViews;
    private List<User> mData;
    private float mBaseElevation;

    public CardPagerAdapterS() {
        mData = new ArrayList<>();
        mViews = new ArrayList<>();
    }

    public void addCardItemS(User user) {
        mViews.add(null);
        mData.add(user);
    }

    public float getBaseElevation() {
        return mBaseElevation;
    }

    @Override
    public CardView getCardViewAt(int position) {
        return mViews.get(position);
    }

    @Override
    public int getCount() {
        return mData.size();
    }

    @Override
    public boolean isViewFromObject(View view, Object object) {
        return view == object;
    }

    @Override
    public Object instantiateItem(ViewGroup container, int position) {
        View view = LayoutInflater.from(container.getContext())
                .inflate(R.layout.card_view_tinder_like, container, false);
        container.addView(view);
        bind(mData.get(position), view);
        CardView cardView = (CardView) view.findViewById(R.id.cardview);

        if (mBaseElevation == 0) {
            mBaseElevation = cardView.getCardElevation();
        }

        cardView.setMaxCardElevation(mBaseElevation * Companion.getMAX_ELEVATION_FACTOR());
        mViews.set(position, cardView);
        return view;
    }

    @Override
    public void destroyItem(ViewGroup container, int position, Object object) {
        container.removeView((View) object);
        mViews.set(position, null);
    }

    private void bind(User user, View view) {
        TextView tvName = view.findViewById(R.id.tv_name_age);
        TextView tvDistance = view.findViewById(R.id.tv_distance);
        ImageView ivAvatar = view.findViewById(R.id.iv_user_avatar);

        tvName.setText(user.getName());
        tvDistance.setText(user.getLocation().toString());
        Transformation transformation = new RoundedCornersTransformation(20, 1);

        RequestOptions requestOptions = new RequestOptions()
                .centerCrop()
                .transforms(transformation);

        RequestBuilder<Drawable> thumbnail = Glide.with(view)
                .load(R.drawable.ic_launcher_background)
                .apply(requestOptions);

        Glide.with(view)
                .load(user.getAvatarUrl())
                .apply(requestOptions)
                .thumbnail(thumbnail)
                .into(ivAvatar);
    }
}

File: app/src/main/java/itis/ru/justtalk/adapters/people/CardPagerAdapterS.kt
package itis.ru.justtalk.adapters.people

import android.annotation.SuppressLint
import android.support.v4.view.PagerAdapter
import android.support.v7.widget.CardView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import itis.ru.justtalk.R
import itis.ru.justtalk.interactor.myprofile.MyProfileInteractor
import itis.ru.justtalk.models.User
import itis.ru.justtalk.repository.UserRepositoryImpl
import itis.ru.justtalk.utils.getDistanceFromLocation
import jp.wasabeef.glide.transformations.RoundedCornersTransformation
import java.util.*
import javax.inject.Inject

class CardPagerAdapterS: PagerAdapter() {
    var myProfileInteractor: MyProfileInteractor = MyProfileInteractor(UserRepositoryImpl(
        FirebaseAuth.getInstance(), FirebaseFirestore.getInstance()))
    private val mViews: MutableList<CardView?>
    val mData: MutableList<User>
    private lateinit var clickListener: (User) -> Unit
    var baseElevation: Float = 0.toFloat()
        private set

    init {
        mData = ArrayList()
        mViews = ArrayList()
    }

    fun setOnClickListener(clickListener: (User) -> Unit) {
        this.clickListener = clickListener
    }

    fun addCardItemS(userList: List<User>) {
        //mViews.add(null)
        mData.addAll(userList)
    }

    override fun getCount(): Int {
        return mData.size
    }

    override fun isViewFromObject(view: View, `object`: Any): Boolean {
        return view === `object`
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val view = LayoutInflater.from(container.context)
            .inflate(R.layout.card_view_tinder_like, container, false)
        container.addView(view)
        bind(mData[position], view)
        val cardView = view.findViewById<View>(R.id.cardview) as CardView

        cardView.setOnClickListener {
            clickListener(mData[position])
        }

        if (baseElevation == 0f) {
            baseElevation = cardView.cardElevation
        }

        cardView.maxCardElevation = baseElevation
        mViews.add(cardView)
        return view
    }

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        container.removeView(`object` as View)
        mViews[position] = null
    }

    @SuppressLint("CheckResult")
    private fun bind(user: User, view: View) {
        val tvName = view.findViewById<TextView>(R.id.tv_name_age)
        val tvDistance = view.findViewById<TextView>(R.id.tv_distance)
        val ivAvatar = view.findViewById<ImageView>(R.id.iv_user_avatar)

        tvName.text = user.name
        myProfileInteractor
            .getMyProfile()
            .subscribe({
                tvDistance.text = getDistanceFromLocation(user.location, it.location)
            }, {
                it.printStackTrace()
            })

        val transformation = RoundedCornersTransformation(20, 1)

        val requestOptions = RequestOptions()
            .centerCrop()
            .transforms(transformation)

        val thumbnail = Glide.with(view)
            .load(R.drawable.image_placeholder)
            .apply(requestOptions)

        Glide.with(view)
            .load(user.avatarUrl)
            .apply(requestOptions)
            .thumbnail(thumbnail)
            .into(ivAvatar)
    }
}
